### PR TITLE
PWN-7341 - Add ethereum kit, and provide keyPair generation repository

### PR DESCRIPTION
**Pull request by Davran331122**
…lic key. Write unit test to validate generating pub key

## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-7341

## Description of Work

Add Ethereum kit, add repository for generate key pair, write unit test for testing generating key pair

### CODE REVIEW

File: ethereumkit/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,77 @@
+plugins {
+    id("org.p2p.wallet.android.library")
+    id("kotlin-kapt")
+}
+
+android {
+    defaultConfig {
+        javaCompileOptions {
```
Comment (gslevinkov) [inline ethereumkit/build.gradle.kts:8]: javaCompileOptions goes double

File: ethereumkit/src/main/java/org/p2p/ethereumkit/api/core/Interfaces.kt [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+import com.google.gson.JsonElement
+import com.google.gson.annotations.SerializedName
+import org.p2p.ethereumkit.api.jsonrpc.JsonRpc
+import io.reactivex.Single
+
+class RpcHandler(val onSuccess: (RpcResponse) -> Unit, val onError: (Throwable) -> Unit)
+typealias SubscriptionHandler = (RpcSubscriptionResponse) -> Unit
+
+data class RpcResponse(val id: Int, val result: JsonElement?, val error: Error?) {
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/api/core/Interfaces.kt:11]: make everything you can as internal to remove autocomplete garbage

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/utils/EthereumUtils.kt [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    private fun cleanHexPrefix(input: String): String {
+        return if (containsHexPrefix(input)) {
+            input.substring(2)
+        } else {
+            input
+        }
+    }
+
+    private fun containsHexPrefix(input: String): Boolean {
+        return input.isNotEmpty() &&
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/utils/EthereumUtils.kt:45]: input.length > 1 already check that input is not empty

File: ethereumkit/src/test/java/org/p2p/ethereumkit/EthereumAddressValidatorTest.kt [hunk 4]
```diff
@@ -0,0 +1,28 @@
+package org.p2p.ethereumkit
+
+import org.junit.Test
+import org.p2p.ethereumkit.external.utils.EthereumUtils
+
+class EthereumAddressValidatorTest {
+
+    @Test
+    fun `validate addresses for Ethereum`() {
```
Comment (gslevinkov) [inline ethereumkit/src/test/java/org/p2p/ethereumkit/EthereumAddressValidatorTest.kt:9]: split into two tests - one for valid addresses and one for invalid.
i suggest given-when-then naming for test naming

File: ethereumkit/src/test/java/org/p2p/ethereumkit/EthereumAddressValidatorTest.kt [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        val address1Valid = "0xff096cc01a7cc98ae3cd401c1d058baf991faf76"
```
Comment (gslevinkov) [inline ethereumkit/src/test/java/org/p2p/ethereumkit/EthereumAddressValidatorTest.kt:10]: can be listOf validAddresses

File: ethereumkit/src/test/java/org/p2p/ethereumkit/EthereumAddressValidatorTest.kt [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+        val address2Valid = "0xff096cc01a7cc98ae3cd401c1d0512af991faf76"
+        val address3Valid = "0xc82d35fcc85c56d9b7c3b1d23e80c49598b2a513"
+        val address4Valid = "0xbed713fdb50a66dd5980cfaad82778236f0dc8df"
+        val address3ValidWithout0x = "ff096cc01a7cc98ae3cd401c1d0512af991faf76"
+        val address1NotValid = "0xff096cc01a7cc98ae3cd401c1d058baf991faf"
+        val address2NotValid = "0xff096cc01123sfwfccc98ae3cd401c1d058baf991faf"
+
+        assert(EthereumUtils.isValidAddress(address1Valid))
```
Comment (gslevinkov) [inline ethereumkit/src/test/java/org/p2p/ethereumkit/EthereumAddressValidatorTest.kt:18]: validAddresses.forEach { 
    val validCheck = EthereumUtils.isValidAddress(address1Valid)
    Assertions.assertThat(validCheck).isTrue
}

File: ethereumkit/src/test/java/org/p2p/ethereumkit/EthereumKeyPairGeneratorTest.kt [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+private const val ETHEREUM_ADDRESS_LENGTH = 42
+
+class EthereumKeyPairGeneratorTest {
+    //This seed phrase was taken from ethereum kit repository, so feel free to use it
+    private val seedPhrase = "apart approve black comfort steel spin real renew tone primary key cherry".split(" ")
+
+    @Before
+    fun setup() {
+        EthereumKit.init()
```
Comment (gslevinkov) [inline ethereumkit/src/test/java/org/p2p/ethereumkit/EthereumKeyPairGeneratorTest.kt:17]: if we need to launch this thing when app starts - better to place it in repository now

File: ethereumkit/src/test/java/org/p2p/ethereumkit/EthereumKeyPairGeneratorTest.kt [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    }
+
+    @Test
+    fun `generate eth key pair by given seed phrase`() {
+        val repository = EthereumKitRepository()
+        val result = repository.generateKeyPair(seedPhrase)
+        assert(result.publicKey.toString().length == ETHEREUM_ADDRESS_LENGTH)
```
Comment (gslevinkov) [inline ethereumkit/src/test/java/org/p2p/ethereumkit/EthereumKeyPairGeneratorTest.kt:24]: Assertions.assertThat(result.publicKey.toString()).hasSize(ETHEREUM_ADDRESS_LENGTH)

**Review by gslevinkov — APPROVED** [review]

**Review by gslevinkov — APPROVED** [review]

**Review by kabramovp2p — APPROVED** [review]

**Pull request merged**